Simplified/AxisService/NYPLContentDownloader.swift

```diff
@@ -0,0 +1,137 @@
+//
+//  NYPLContentDownloader.swift
+//  Simplified
+//
+//  Created by Raman Singh on 2021-04-22.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+
+protocol NYPLAxisContentDownloading {
+  func downloadItem(from url: URL,
+                    _ completion: @escaping (Result<Data, Error>) -> Void)
+}
+
+
+/// A content downloader which stops downloading content upon receiving a failure.
+class NYPLContentDownloader: NYPLAxisContentDownloading {
+  
+  private var tasks: ThreadSafeDictionaryContainer<URL, URLSessionDataTask>
+  private var handlers: ThreadSafeDictionaryContainer<URL, (Result<Data, Error>) -> Void>
+  private let errored: ThreadSafeValueContainer<Bool>
+  let networkExecutor: NYPLAxisNetworkExecuting
+  
+  init(networkExecuting: NYPLAxisNetworkExecuting) {
+    self.tasks = ThreadSafeDictionaryContainer()
+    self.handlers = ThreadSafeDictionaryContainer()
+    self.errored = ThreadSafeValueContainer(value: false)
+    self.networkExecutor = networkExecuting
+  }
+  
+  func downloadItem(from url: URL,
+                    _ completion: @escaping (Result<Data, Error>) -> Void) {
+    
+    // We dont want to download anything if it's already being downloaded or if
+    // we experienced an error downloading something before.
+    
+    guard !errored.value, tasks[url] == nil, handlers[url] == nil else {
+      return
+    }
+    
+    handlers[url] = completion
+    tasks[url] = self.getTask(for: url)
+  }
+  
+  private func getTask(for url: URL) -> URLSessionDataTask {
+    
+    let request = URLRequest(url: url,
+                             cachePolicy: .reloadIgnoringLocalCacheData,
+                             timeoutInterval: networkExecutor.requestTimeout)
+    
+    let task = networkExecutor.GET(request) { [weak self] (result) in
+      guard let self = self else { return }
+      switch result {
+      case .success(let data):
+        self.handlers[url]?(.success(data))
+        self.tasks[url] = nil
+        self.handlers[url] = nil
+      case .failure(let error):
+        // Here we perform short-circuiting so that we don't download anything
+        // else after receiving a failure since a book with a missing file is
+        // unusable.
+        self.handlers[url]?(.failure(error))
+        self.tasks.forEach {
+          $1.cancel()
+        }
+        self.tasks.removeAll()
+        self.handlers.removeAll()
+        self.errored.value = true
+      }
+    }
+    
+    return task
+  }
+  
+}
+
+private class ThreadSafeDictionaryContainer<S: Hashable,T> {
+  
+  private var dictionary: [S:T]
+  private let queue = DispatchQueue(label: "AxixThreadSafeDictionaryContainer")
```

Comment:
minor: typo - Axis, also on the one below

Reply:
haha great catch thanks